overlord/hookstate/ctlcmd/mount.go

```diff
@@ -0,0 +1,219 @@
+		return false
+	}
+
+	if !matchPathAttribute(m.Positional.Where, attributes["where"], m.snapInfo) {
+		return false
+	}
+
+	if m.Type != "" {
+		// TODO: what should we do if the type is not given?
+		if types, ok := attributes["type"].([]interface{}); ok {
+			found := false
+			for _, iface := range types {
+				if typeString, ok := iface.(string); ok && typeString == m.Type {
+					found = true
+					break
+				}
+			}
+			if !found {
+				return false
+			}
+		} else {
+			return false
+		}
+	}
+
+	if optionsIfaces, ok := attributes["options"].([]interface{}); ok {
+		var allowedOptions []string
+		for _, iface := range optionsIfaces {
+			if option, ok := iface.(string); ok {
+				allowedOptions = append(allowedOptions, option)
+			}
+		}
+		for _, option := range m.optionsList {
+			if !strutil.ListContains(allowedOptions, option) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// checkConnections checks whether the established connections give the snap
+// permission to execute the mount command
+func (m *mountCommand) checkConnections(context *hookstate.Context) error {
+	snapName := context.InstanceName()
+
+	st := context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	conns, err := ifacestate.ConnectionStates(st)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get connections: %s", err)
+	}
+
+	m.snapInfo, err = snapstate.CurrentInfo(st, snapName)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get snap info: %s", err)
+	}
+
+	for connId, connState := range conns {
+		if connState.Interface != "mount-control" {
+			continue
+		}
+
+		if connState.Undesired || connState.HotplugGone {
+			continue
+		}
+
+		connRef, err := interfaces.ParseConnRef(connId)
+		if err != nil {
+			return err
+		}
+
+		if connRef.PlugRef.Snap != snapName {
+			continue
+		}
+
+		mounts, ok := connState.StaticPlugAttrs["mount"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		for _, mountAttributes := range mounts {
+			if m.matchConnection(mountAttributes.(map[string]interface{})) {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("no matching mount-control connection found")
+}
+
+func (m *mountCommand) createMountUnit(sysd systemd.Systemd) (string, error) {
+	snapName := m.snapInfo.InstanceName()
+	revision := m.snapInfo.SnapRevision().String()
+	lifetime := systemd.TransientUnit
+	if m.Persistent {
+		lifetime = systemd.PersistentUnit
+	}
+	return sysd.AddMountUnitFileWithOptions(&systemd.MountUnitOptions{
+		Lifetime: lifetime,
+		SnapName: snapName,
+		Revision: revision,
+		What:     m.Positional.What,
+		Where:    m.Positional.Where,
+		Fstype:   m.Type,
+		Options:  m.optionsList,
+		Creator:  "mount-control",
+	})
+}
+
+func (m *mountCommand) Execute([]string) error {
+	fmt.Printf("Executing mount command %q -> %q, persistent: %v\n", m.Positional.What, m.Positional.Where, m.Persistent)
+
+	context := m.context()
+	if context == nil {
+		return fmt.Errorf("cannot mount without a context")
```

Comment:
this error message isn't super clear to the user, but in general I think our snapctl error messages around context's being missing are today unclear so it's probably fine

Reply:
It looks to me that all the commands have a similar error "cannot ... without a context" :-)

Would it be a good idea to change

    func (c *baseCommand) context() *hookstate.Context {
        return c.c
    }

to

    func (c *baseCommand) context() (*hookstate.Context, error) {
        if c.c != nil {
            return c.c, nil
        } else {
            return nil, fmt.Errorf("The snapctl command can only be used from inside a snap")
        }
    }

and then change all the callers accordingly, or are there cases where one can use snapctl without a context?

Reply:
AFAIK, snapctl is only meant to be used from inside a snap or inside a hook, so it should always have a context, the one exception being `snapctl --help` I guess